### PR TITLE
feat: CNS-949: add feegrant tx to stake cli

### PR DESCRIPTION
**Pull request by oren-lava**
In this PR, I implemented a new flag for the `stake-provider` CLI command called `--grant-provider-gas-fees-auth` which allows the provider address (that operates the provider process) use the vault address' funds to fund gas fees for provider-related TXs.

The funded TXs: `relay-payment`, `freeze`, `unfreeze`, `modify-provider`, `detection` (conflict module), `conflict-vote-commit` and `conflict-vote-reveal`.

Tip: when using the provider address to run one of the TXs above, use Cosmos' `--fee-granter` flag to let the vault pay for gas fees (without it, the gas won't be funded, even if you used the `--grant-provider-gas-fees-auth` flag when staking the provider).

### CODE REVIEW

File: x/pairing/client/cli/tx_stake_provider.go [hunk 1]
```diff
@@ -365,3 +400,58 @@ func getValidator(clientCtx client.Context, provider string) string {
...
+
+	grant, err := feegrant.NewAllowedMsgAllowance(&feegrant.BasicAllowance{}, []string{
+		// pairing module TXs
+		sdk.MsgTypeURL(&types.MsgRelayPayment{}),
+		sdk.MsgTypeURL(&types.MsgStakeProvider{}), // for modify-provider TX
+		sdk.MsgTypeURL(&types.MsgFreezeProvider{}),
+		sdk.MsgTypeURL(&types.MsgUnfreezeProvider{}),
+
+		// conflict module TXs
+		sdk.MsgTypeURL(&conflicttypes.MsgDetection{}),
```
Comment (Yaroms) [inline x/pairing/client/cli/tx_stake_provider.go:434]: this is done b the consumer, it is not needed

Comment (omerlavanet) [inline x/pairing/client/cli/tx_stake_provider.go:434]: detection is a consumer

File: x/pairing/client/cli/tx_unstake_provider.go [hunk 2]
```diff
@@ -54,6 +58,11 @@ func CmdUnstakeProvider() *cobra.Command {
 					return err
 				}
 				msgs = append(msgs, msg)
+
+				revokeGrantFeeMsg := CreateRevokeFeeGrantMsg(clientCtx, chainID)
```
Comment (Yaroms) [inline x/pairing/client/cli/tx_unstake_provider.go:62]: do we need to check if there is a feegrant at all?

Comment (omerlavanet) [inline x/pairing/client/cli/tx_unstake_provider.go:62]: we do not check it exists, it can fail the unstake too if the feegrant fails

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/pairing/client/cli/tx_stake_provider.go [hunk 3]
```diff
@@ -365,3 +400,58 @@ func getValidator(clientCtx client.Context, provider string) string {
...
+		sdk.MsgTypeURL(&conflicttypes.MsgConflictVoteCommit{}),
+		sdk.MsgTypeURL(&conflicttypes.MsgConflictVoteReveal{}),
+	})
+	if err != nil {
+		utils.LavaFormatError("failed granting feegrant for gas fees", err,
+			utils.LogAttr("granter", granter),
+			utils.LogAttr("grantee", grantee),
+		)
+		return nil
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/client/cli/tx_stake_provider.go]: done. see https://github.com/lavanet/lava/commit/165e856d35ac919bd67830cb66b0ccc4f8b156e6

File: x/pairing/client/cli/tx_stake_provider.go (hunk 1 above)
Comment (oren-lava) [reply x/pairing/client/cli/tx_stake_provider.go]: done. see https://github.com/lavanet/lava/commit/165e856d35ac919bd67830cb66b0ccc4f8b156e6

Comment (oren-lava) [reply x/pairing/client/cli/tx_stake_provider.go]: done. see https://github.com/lavanet/lava/commit/165e856d35ac919bd67830cb66b0ccc4f8b156e6

File: x/pairing/client/cli/tx_unstake_provider.go (hunk 2 above)
Comment (oren-lava) [reply x/pairing/client/cli/tx_unstake_provider.go]: done. see https://github.com/lavanet/lava/commit/af338548b5293fffb5736dfc1b4c246893ce6c0b

Comment (oren-lava) [reply x/pairing/client/cli/tx_unstake_provider.go]: done. see https://github.com/lavanet/lava/commit/af338548b5293fffb5736dfc1b4c246893ce6c0b

File: protocol/statetracker/tx_sender.go [hunk 4]
```diff
@@ -318,31 +332,128 @@ func NewProviderTxSender(ctx context.Context, clientCtx client.Context, txFactor
...
+			// Allowance doesn't return an error if allowance doesn't exist. Error is returned on parsing issues
+			utils.LavaFormatWarning("could not get gas fee allowance for provider and vault", err,
+				utils.LogAttr("provider", pts.clientCtx.FromAddress.String()),
+				utils.LogAttr("vault", stakeEntry.Vault),
+				utils.LogAttr("chain_id", stakeEntry.Chain),
+			)
+			continue
+		}
+		if res.Allowance != nil {
+			vault.isFeeGranter = true
```
Comment (Yaroms) [inline protocol/statetracker/tx_sender.go:396]: shouldnt we just not keep it in the map if it is not a fee granter?

**Review by Yaroms — APPROVED** [review]

**Review by omerlavanet — APPROVED** [review]
Very nice!!!

**Pull request merged**